73-offline-attestation.md

```diff
@@ -0,0 +1,302 @@
+node. Furthermore we cannot even rely on the accessibility (or even the
+existence) of the server-side components of the Keylime cluster (i.e.,
+`registrar` and `verifier`) and thus need to design with these boundary
+conditions in mind.  
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Add functionality on the `registrar` to record (in a tamper-resistant
+  transparency log) the association between the EK and AIK (i.e.
+`tpm2_makecredential`)
+- Add functionality on the `verifier` to record (in a time-series persistent
+  datastore) all the information needed to perform attestation standlone (i.e.,
+quotes and MB/IMA logs)
+- Add a new CLI which will interface with the aforementioned persistent stores,
+  and will call the main, umodified `verifier` code in order to do post-facto
+attestation.  
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- The interaction between the time-series persistent datastore and
+  tamper-resistant transparency log will be done by keylime user/operator.
+Inside the core Keylime, a "plugin" architecture will be adopted (very much
+like the "policies" for Measured Boot) and the implementation details of the
+code which will interact with such stores are outside of the scope.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- The `registrar` will be modified to, upon initial `agent` registration -
+  which includes the execution of `tpm2_makecredential` - record the EK, AIK
+into a JSON file, sign it (using the private key generated as part of the
+certificates for mTLS interaction with both `tenant` and `registrar`) and then
+make a record of it on a tamper-resistant transparency log (e.g., Rekor). In
+addition to that, it will store the JSON file, the signature, and the public
+key on the time-series persistent datastore. This should allow an external
+component/user to check, provided that there is trust on the `registrar`, that
+a particular AIK is indeed tied to a particular EK. The reason for having this
+data stored into a time-series is due to the fact that AIKs are regenereated
+every time an `agent` is restarted on Keylime.
+- The `verifier` will be modified to take the `json_response` (python
+  dictionary) from the `agent` - which will include both quotes and logs (MB
+and IMA) - `agent` data (python dictionary) from the SQL database (internal to
+Keylime) and the `agentAttestState` python object, combine it into a single
+record and store it on the time-series persistent datastore. 
```

Comment:
IIUC this proposal will store marshaled Python objects? IMHO we should make the effort (if we indeed decide to store this data) to serialize in a more agnostic format (json, yaml).  So a 3rd party software can still read it and replicate the attestation.

Reply:
I will update it :-) I've studied the code further and simple `json` is enough, no need to marshal/pickle anymore.